src/plugins/core/conditional_format.ts

```diff
@@ -58,37 +58,95 @@ export class ConditionalFormatPlugin
-          case "CHANGE":
+  adaptCFFormulas(applyChange: ApplyRangeChange) {
+    for (const sheetId in this.cfRules) {
+      for (const rule of this.cfRules[sheetId]) {
+        if (rule.rule.type === "DataBarRule" && rule.rule.rangeValues) {
+          const change = applyChange(rule.rule.rangeValues);
+          switch (change.changeType) {
+            case "REMOVE":
+              this.history.update(
+                "cfRules",
+                sheetId,
+                this.cfRules[sheetId].indexOf(rule),
+                "rule",
+                //@ts-expect-error
+                "rangeValues",
+                undefined
+              );
+              break;
+            case "RESIZE":
+            case "MOVE":
+            case "CHANGE":
+              this.history.update(
+                "cfRules",
+                sheetId,
+                this.cfRules[sheetId].indexOf(rule),
+                "rule",
+                //@ts-expect-error
+                "rangeValues",
+                change.range
+              );
+              break;
+          }
+        } else if (rule.rule.type === "CellIsRule") {
+          for (let i = 0; i < rule.rule.values.length; i++) {
             this.history.update(
               "cfRules",
               sheetId,
               this.cfRules[sheetId].indexOf(rule),
               "rule",
               //@ts-expect-error
-              "rangeValues",
-              change.range
+              "values",
+              i,
+              this.getters.adaptFormulaStringDependencies(sheetId, rule.rule.values[i], applyChange)
             );
-            break;
+          }
+        } else if (rule.rule.type === "IconSetRule") {
+          for (const inflectionPoint of ["lowerInflectionPoint", "upperInflectionPoint"] as const) {
+            if (rule.rule[inflectionPoint].type === "formula") {
+              this.history.update(
+                "cfRules",
+                sheetId,
+                this.cfRules[sheetId].indexOf(rule),
+                "rule",
+                //@ts-expect-error
+                inflectionPoint,
+                "value",
+                this.getters.adaptFormulaStringDependencies(
+                  sheetId,
+                  rule.rule[inflectionPoint].value,
+                  applyChange
+                )
+              );
+            }
+          }
+        } else if (rule.rule.type === "ColorScaleRule") {
+          for (const value of ["minimum", "maximum", "midpoint"] as const) {
+            if (rule.rule[value]?.type === "formula" && rule.rule[value]?.value) {
+              this.history.update(
+                "cfRules",
+                sheetId,
+                this.cfRules[sheetId].indexOf(rule),
+                "rule",
+                //@ts-expect-error
+                value,
+                "value",
+                this.getters.adaptFormulaStringDependencies(
+                  sheetId,
+                  rule.rule[value]!.value!,
```

Comment:
coud you move `rule.rule[value]!.value!` to a variable before the if statement and use it here and in the if condition.
It'll allow to get rid of `!`

Reply:
value is only present based on the type, having the const would then mean i have to split the if condition in 2.
I feel like ! is clearer than two nested ifs with a const in between them